libs/driver/src/VideoDriver.cpp

```diff
@@ -73,4 +76,60 @@ void VideoDriver::SetNewSize(VideoMode windowSize, Extent renderSize)
 {
     windowSize_ = windowSize;
     renderSize_ = renderSize;
+
+    const auto ratioXY = PointF(renderSize_) / PointF(windowSize_.width, windowSize_.height);
+    dpiScale_ = (ratioXY.x + ratioXY.y) / 2.f; // use the average ratio of both axes
+
+    if(autoGuiScale_)
+    {
+        unsigned percent;
+        getGuiScaleRange(nullptr, nullptr, &percent);
+        guiScale_ = GuiScale(percent);
+    }
+
+    scaledRenderSize_ = guiScale_.screenToView<Extent>(renderSize);
+}
+
+void VideoDriver::setGuiScalePercent(unsigned percent)
+{
+    autoGuiScale_ = (percent == 0);
+    if(autoGuiScale_)
+        getGuiScaleRange(nullptr, nullptr, &percent);
+
+    if(guiScale_.percent() == percent)
+        return;
+
+    // translate current mouse position to screen space
+    const auto screenPos = guiScale_.viewToScreen(mouse_xy.pos);
+
+    guiScale_ = GuiScale(percent);
+    scaledRenderSize_ = guiScale_.screenToView<Extent>(renderSize_);
+    CallBack->WindowResized();
+
+    // move cursor to new position in view space
+    // must happen after window resize event to avoid drawing spurious hover events
+    mouse_xy.pos = guiScale_.screenToView(screenPos);
+    CallBack->Msg_MouseMove(mouse_xy);
+}
+
+void VideoDriver::getGuiScaleRange(unsigned* minPercent, unsigned* maxPercent, unsigned* recommendedPercent) const
```

Comment:
I was checking this noting that I didn't fully understood it initially. I'd rather have this return a simple struct containing min/max/recommended percent than out-parameters passed as pointers where you don't know if the original values matter until you read the code twice.

Maybe even split out the recommended part into an extra function which would even simplify https://github.com/Return-To-The-Roots/s25client/pull/1594/files#diff-334108d4fb1ef8a8a66c17c2706fc8f30d515828a6d739e53d2e819d842835ddR595-R598

Reply:
This OK?

```cpp
// in GuiScale.h
struct GuiScaleRange {
    unsigned minPercent; // wanted to include the unit somewhere and picked the struct members
    unsigned maxPercent;
    unsigned recommendedPercent;
};

// range function is part of the interface
virtual GuiScaleRange VideoInterface::getGuiScaleRange() const;

// additional convenience function only exists in the wrapper
unsigned VideoDriverWrapper::getRecommendedGuiScalePercent() const {
    return driver->getGuiScaleRange().recommendedPercent;
}
```

Reply:
Yeah looks good!   
This way I'd even omit the `getRecommendedGuiScalePercent` as it doesn't provide  much benefit if `getGuiScaleRange` has it has a member.